cpp/src/binary/jit/code/kernel.cpp

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright 2018-2019 BlazingDB, Inc.
+ *     Copyright 2018 Christian Noboa Mardini <christian@blazingdb.com>
+ *     Copyright 2018 Rommel Quintanilla <rommel@blazingdb.com>
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace gdf {
+namespace binops {
+namespace jit {
+namespace code {
+
+const char* kernel =
+R"***(
+    #include <cstdint>
+    #include "traits.h"
+    #include "operation.h"
+
+    template <typename TypeOut, typename TypeVax, typename TypeVay, typename TypeOpe>
+    __global__
+    void kernel_v_s(int size,
+                    TypeOut* out_data, TypeVax* vax_data, TypeVay* vay_data,
+                    uint32_t* out_valid, uint32_t* vax_valid) {
+        int tid = threadIdx.x;
+        int blkid = blockIdx.x;
+        int blksz = blockDim.x;
+        int gridsz = gridDim.x;
+
+        int start = tid + blkid * blksz;
+        int step = blksz * gridsz;
+
+        for (int i=start; i<size; i+=step) {
+            out_data[i] = TypeOpe::template operate<TypeOut, TypeVax, TypeVay>(vax_data[i], vay_data[0]);
+
+            if ((out_valid != nullptr) && (i % warpSize) == 0) {
+                int index = i / warpSize;
+                uint32_t vax_valid_word = (vax_valid != nullptr) ? vax_valid[index] : 0xffffffff;
+                out_valid[index] = vax_valid_word;
+            }
+        }
+    }
+
+    template <typename TypeOut, typename TypeVax, typename TypeVay, typename TypeOpe>
+    __global__
+    void kernel_v_v(int size,
+                    TypeOut* out_data, TypeVax* vax_data, TypeVay* vay_data,
+                    uint32_t* out_valid, uint32_t* vax_valid, uint32_t* vay_valid) {
+        int tid = threadIdx.x;
+        int blkid = blockIdx.x;
+        int blksz = blockDim.x;
+        int gridsz = gridDim.x;
+
+        int start = tid + blkid * blksz;
+        int step = blksz * gridsz;
+
+        for (int i=start; i<size; i+=step) {
+            out_data[i] = TypeOpe::template operate<TypeOut, TypeVax, TypeVay>(vax_data[i], vay_data[i]);
+
+            if ((out_valid != nullptr) && (i % warpSize) == 0) {
```

Comment:
The way the current binops code does it is it completely does away with bit mask calculation in the main api call and adds another api to just do a bit mask AND. It leaves the decision to AND the masks to the API user. Should I just remove the bit mask calculation altogether?

Reply:
I'm not sure I understand.

I do feel that the bitmask handling code should be completely removed from the two binary function kernels. It can be handled outside of the kernel much more efficiently and cleanly via memcpys

Reply:
By current binops code, I meant the compiled implementation [here](https://github.com/rapidsai/cudf/blob/6a637dc12d3e382a278378c3ed5841d3def62684/cpp/src/binary/binary_ops.cu#L519) rather than the JIT version. And I agree with your efficiency argument, my question was, should a call to `gdf_binary_operation_[v_v/v_s/s_v]` also do the bit mask calculation for `out` or should this be a separate api call?

Reply:
If someone calls `gdf_binary_operation*`, then the bitmask for the output column should be updated for them appropriately. It should not require the user to take any further action. 

Furthermore, I notice that the existing homogeneous type binary operation also touches the bitmask. 

https://github.com/rapidsai/cudf/blob/6a637dc12d3e382a278378c3ed5841d3def62684/cpp/src/binary/binary_ops.cu#L21

There's no reason for this. You can just operate on whatever garbage values are in memory for the null values, and then set the bitmask appropriately. 

The bottom line is this: no matter if a binary operation is taking the static path or the JIT path, there should only be a single code path for updating the bitmask for the output column, and it should be done _outside_ of the kernel doing the actual `lhs op rhs`.

Reply:
I noticed that too and what's worse is that this special checking for null values is actually pretty expensive. I benchmarked and found that adding 10mil columns took 3 ms without the checking and 14 ms with the checking.